sql: more assorted telemetry fixes

knz opened the pull request:
Fixes #31990. Requested by @awoods187.

knz committed:
sql: introduce references to unimplemented #32005

Release note: None

knz committed:
sql: make ALTER CONSTRAINT reference #31632

Release note: None

knz committed:
sql: reference #32555 for ALTER TABLE RENAME CONSTRAINT

Release note: None

knz committed:
sql: reference #9851 for ALTER SET TYPE + detail conversion

The telemetry data did not include the requested type conversion. This
patch adds it.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32556)
<!-- Reviewable:end -->

knz committed:
sql: do not ignore the WHERE clause after INSERT ON CONFLICT

Release note (bug fix): CockroachDB previously silently and
incorrectly accepted then ignored a WHERE clause after INSERT ON
CONFLICT. This is now properly recognized and will produce an error
reporting the feature is unimplemented.

knz committed:
sql: reference #28161 for INSERT ON CONFLICT ON CONSTRAINT

Release note: None

knz committed:
sql: reference #20305 for FK composite match specifications

Release note: None

knz committed:
importccl: reference #32560, #32559 and #32561 for unimplemented types

Release note: None

knz committed:
sql: reference #29639 for ALTER TABLE ADD CONSTRAINT CHECK ...

Release note: None

knz committed:
sql: reference #32562 for SET LOCAL

Release note: None

knz committed:
sql: reference #8855 for ALTER TABLE ... ADD CONSTRAINT ... REFERENCES

Release note: None

knz committed:
sql: reference #32563 for COLLATION FOR

Release note: None

knz committed:
sql: detail the type for unsupported casts using "T() SCONST"

Release note: None

knz committed:
sql: reference #32564 for INTERVAL with precision

Release note: None

knz committed:
sql: reference #32098 and #32565 for time/timestamp with precision

This also somewhat simplifies the grammar.

Release note: None

knz committed:
sql: reference #20961 and #32567 during planning of sequence options

Release note: None

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

with some minor comments

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 3 of 3 files at r3, 2 of 2 files at r4, 8 of 8 files at r5, 2 of 2 files at r6, 2 of 2 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, 2 of 2 files at r10, 1 of 1 files at r11, 2 of 2 files at r12, 2 of 2 files at r13, 3 of 3 files at r14, 3 of 3 files at r15, 1 of 1 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32556#-:-LSBreZJ7ULeRCnLFhGn:bt86nmq)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/ccl/importccl/read_import_mysql.go, line 614 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32556#-LSBdFqsAr8uBHO-8ZDs:-LSBdFqt1vocGkMZgoXg:bs8vitx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d0f348655c093f4d94709df8b1973815bd43bc0/pkg/ccl/importccl/read_import_mysql.go#L614)):*
> ```Go
> 		def.Type = coltypes.JSON
> 
> 	case mysqltypes.Set:
> ```

Is there a test for these?
___
*[pkg/sql/sequence.go, line 245 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/32556#-LSBrZhU5RzccoHLY-vF:-LSBrZhU5RzccoHLY-vG:b4u25rr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1fd2dcaf625ac62354d33937cd97f6c33cdacaa/pkg/sql/sequence.go#L245)):*
> ```Go
> 				// Do nothing; this is the default.
> 			case v > 1:
> 				return pgerror.UnimplementedWithIssueErrorf(32567,
> ```

tests for these?
___
*[pkg/sql/parser/sql.y, line 6812 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/32556#-LSBrQQr6HVZTgK_GrlB:-LSBrQQr6HVZTgK_GrlC:b684qbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5aa4ea9497ea019589f91edec2de7e0b6aec9d3/pkg/sql/parser/sql.y#L6812)):*
> ```Yacc
> 
> opt_timezone:
>   WITH_LA TIME ZONE	{ $$.val = true; }
> ```

leftover tabs


<!-- Sent from Reviewable.io -->

mjibson commented on pkg/sql/parser/sql.y:
```diff
@@ -6764,38 +6785,34 @@ const_datetime:
+opt_timezone:
+  WITH_LA TIME ZONE	{ $$.val = true; }
+| WITHOUT TIME ZONE	{ $$.val = false; }
+| /*EMPTY*/			{ $$.val = false; }
```
nit: whitespaces appear incorrect here

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32556)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/ccl/importccl/read_import_mysql.go, line 614 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/32556#-LSBdFqsAr8uBHO-8ZDs:-LSFTBCqAlMZxzaMBRXa:bz4ww2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d0f348655c093f4d94709df8b1973815bd43bc0/pkg/ccl/importccl/read_import_mysql.go#L614)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is there a test for these?
</blockquote></details>

Filed #32592 to follow up.
___
*[pkg/sql/sequence.go, line 245 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/32556#-LSBrZhU5RzccoHLY-vF:-LSFR_2q4dAitsij71iy:bpfua9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1fd2dcaf625ac62354d33937cd97f6c33cdacaa/pkg/sql/sequence.go#L245)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

tests for these?
</blockquote></details>

Yes they were present previously already in `logic_test/sequences`: see `statement error pgcode 0A000 ...`
___
*[pkg/sql/parser/sql.y, line 6812 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/32556#-LSBrQQr6HVZTgK_GrlB:-LSFRsHlE2NzOPvZ8SmR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5aa4ea9497ea019589f91edec2de7e0b6aec9d3/pkg/sql/parser/sql.y#L6812)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

leftover tabs
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz commented:
TFYRs!

bors r+
